File: system-tests/junit-run-listener-test/src/test/java/com/teamscale/tia/client/JUnitRunListenerSystemTest.java
package com.teamscale.tia.client;

import com.teamscale.report.testwise.model.ETestExecutionResult;
import com.teamscale.report.testwise.model.TestwiseCoverageReport;
import com.teamscale.test.commons.SystemTestUtils;
import com.teamscale.test.commons.TeamscaleMockServer;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

/**
 * Runs several Maven projects' Surefire tests that have the agent attached and one of our JUnit run listeners enabled.
 * Checks that this produces a correct coverage report.
 */
public class JUnitRunListenerSystemTest {

	private static TeamscaleMockServer teamscaleMockServer = null;

	@BeforeEach
	public void startFakeTeamscaleServer() throws Exception {
		if (teamscaleMockServer == null) {
			teamscaleMockServer = new TeamscaleMockServer(SystemTestUtils.TEAMSCALE_PORT).acceptingReportUploads();
		}
		teamscaleMockServer.reset();
	}

	/** Tests the JUnit 5 TestExecutionListener. */
	@Test
	public void testJUnit5TestExecutionListener() throws Exception {
		SystemTestUtils.runMavenTests("junit5-maven-project");

		TestwiseCoverageReport report = teamscaleMockServer.getOnlyTestwiseCoverageReport("part");
		assertThat(report.tests).hasSize(2);
		assertAll(() -> {
			assertThat(report.tests).extracting(test -> test.uniformPath)
					.containsExactlyInAnyOrder("JUnit4ExecutedWithJUnit5Test/testAdd()", "JUnit5Test/testAdd()");
			assertThat(report.tests).extracting(test -> test.result)
					.containsExactlyInAnyOrder(ETestExecutionResult.PASSED, ETestExecutionResult.PASSED);
			assertThat(report.tests).extracting(SystemTestUtils::getCoverageString)
					.containsExactly("SystemUnderTest.java:3,6", "SystemUnderTest.java:3,6");
		});
	}

	/** Tests the JUnit 4 RunListener. */
	@Test
	public void testJUnit4RunListener() throws Exception {
		SystemTestUtils.runMavenTests("junit4-maven-project");

		TestwiseCoverageReport report = teamscaleMockServer.getOnlyTestwiseCoverageReport("part");
		assertThat(report.tests).hasSize(1);
		assertAll(() -> {
			assertThat(report.tests).extracting(test -> test.uniformPath)
					.containsExactlyInAnyOrder("JUnit4Test/testAdd");
			assertThat(report.tests).extracting(test -> test.result)
					.containsExactlyInAnyOrder(ETestExecutionResult.PASSED);
			assertThat(report.tests).extracting(SystemTestUtils::getCoverageString)
					.containsExactly("SystemUnderTest.java:3,6");
		});
	}

}

File: system-tests/junit-run-listener-test/src/test/kotlin/com/teamscale/tia/client/JUnitRunListenerSystemTest.kt
package com.teamscale.tia.client

import com.teamscale.report.testwise.model.ETestExecutionResult
import com.teamscale.report.testwise.model.TestInfo
import com.teamscale.report.testwise.model.TestwiseCoverageReport
import com.teamscale.test.commons.SystemTestUtils
import com.teamscale.test.commons.SystemTestUtils.coverage
import com.teamscale.test.commons.SystemTestUtils.runMavenTests
import com.teamscale.test.commons.TeamscaleMockServer
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.iterable.ThrowingExtractor
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll
import org.junit.jupiter.api.function.Executable

/**
 * Runs several Maven projects' Surefire tests that have the agent attached and one of our JUnit run listeners enabled.
 * Checks that this produces a correct coverage report.
 */
class JUnitRunListenerSystemTest {
	@BeforeEach
	@Throws(Exception::class)
	fun startFakeTeamscaleServer() {
		if (teamscaleMockServer == null) {
			teamscaleMockServer = TeamscaleMockServer(SystemTestUtils.TEAMSCALE_PORT).acceptingReportUploads()
		}
		teamscaleMockServer?.reset()
	}

	/** Tests the JUnit 5 TestExecutionListener.  */
	@Test
	@Throws(Exception::class)
	fun testJUnit5TestExecutionListener() {
		runMavenTests("junit5-maven-project")

		val report: TestwiseCoverageReport = teamscaleMockServer!!.getOnlyTestwiseCoverageReport("part")
		assertThat(report.tests).hasSize(2)
		assertAll({
			assertThat(report.tests)
				.extracting<String> { it.uniformPath }
				.containsExactlyInAnyOrder("JUnit4ExecutedWithJUnit5Test/testAdd()", "JUnit5Test/testAdd()")
			assertThat(report.tests)
				.extracting<ETestExecutionResult> { it.result }
				.containsExactlyInAnyOrder(ETestExecutionResult.PASSED, ETestExecutionResult.PASSED)
			assertThat(report.tests)
				.extracting<String> { it.coverage }
				.containsExactly("SystemUnderTest.java:3,6", "SystemUnderTest.java:3,6")
		})
	}

	/** Tests the JUnit 4 RunListener.  */
	@Test
	@Throws(Exception::class)
	fun testJUnit4RunListener() {
		runMavenTests("junit4-maven-project")

		val report = teamscaleMockServer!!.getOnlyTestwiseCoverageReport("part")
		assertThat(report.tests).hasSize(1)
		assertAll({
			Assertions.assertThat(report.tests)
				.extracting<String> { it.uniformPath }
				.containsExactlyInAnyOrder("JUnit4Test/testAdd")
			Assertions.assertThat(report.tests)
				.extracting<ETestExecutionResult> { it.result }
				.containsExactlyInAnyOrder(ETestExecutionResult.PASSED)
			Assertions.assertThat(report.tests)
				.extracting<String> { it.coverage }
				.containsExactly("SystemUnderTest.java:3,6")
		})
	}

	companion object {
		private var teamscaleMockServer: TeamscaleMockServer? = null
	}
}
